### PR TITLE
IAM Roles

### PR DESCRIPTION
**Pull request by dpetzold**
- Dropped name argument from create_policy. Is was being called by create_template without any arguments.                                                                                                                                                                                                                                     
- Check at least one Ec2Role or LambdaRole is defined.                                                                                                                                                                                                                      
- Added support for new variable AttachedPolicies. If defined its values are passed to ManagedPolicyArns in the IAM Role create.
- Added new variable InstanceProfile. If true the role will be created as an instance profile.
- Added new variable Path. If defined the Path will be set on the Role and the InstanceProfile.
- Coverage support. Run coverage with tests                                                                                                                                                                                                                                 
- Added tests.

### CODE REVIEW

File: stacker_blueprints/iam_roles.py [hunk 1]
```diff
@@ -36,21 +52,60 @@ def __init__(self, *args, **kwargs):
...
     def create_role(self, name, assumerole_policy):
         t = self.template
+        v = self.get_variables()
+
+        role_kwargs = {
+            'AssumeRolePolicyDocument': assumerole_policy,
+        }
+
+        attached_policies = v['AttachedPolicies']
+        if attached_policies and len(attached_policies):
```
Comment (phobologic) [inline stacker_blueprints/iam_roles.py]: You shouldn't need the `len` test here, since anything that's length 0 (a empty list, string, dictionary) will evaluate as False in python.  You could replace these two lines with:

```
role_kwargs["ManagedPolicyArns"] = attached_policies or NoValue
```

(You'll want to import `NoValue` from troposphere at the top of the file if it's not already)

File: stacker_blueprints/iam_roles.py [hunk 2]
```diff
@@ -36,21 +52,60 @@ def __init__(self, *args, **kwargs):
...
+            role_kwargs['ManagedPolicyArns'] = attached_policies
+
+        path = v['Path']
```
Comment (phobologic) [inline stacker_blueprints/iam_roles.py]: Same basic concept as above using `NoValue` should work here as well.

File: stacker_blueprints/iam_roles.py [hunk 3]
```diff
@@ -36,21 +52,60 @@ def __init__(self, *args, **kwargs):
...
 
         t.add_output(
             Output(name + "RoleName", Value=Ref(role))
         )
+
         t.add_output(
             Output(name + "RoleArn", Value=GetAtt(role.title, "Arn"))
         )
 
+        if v['InstanceProfile']:
```
Comment (phobologic) [inline stacker_blueprints/iam_roles.py]: This won't work the way you want, I don't believe, since it's creating an `InstanceProfile` PER `Role` you create, but it's naming the `InstanceProfile` output the same each time.

Is the intent to create an instance profile per role, or one instance profile with all of the roles?

Also, is it worth checking that the role is an ec2 role, and not a Lambda role, before creating the profile?

File: stacker_blueprints/iam_roles.py [hunk 4]
```diff
@@ -92,17 +147,24 @@ def create_policy(self, name):
         )
 
         t.add_output(
-            Output(name + "PolicyName", Value=Ref(policy))
+            Output("PolicyName", Value=Ref(policy))
```
Comment (phobologic) [inline stacker_blueprints/iam_roles.py]: This could be a breaking change for people that use this blueprint already, right?  If this output was relied on in another stack, that stack will fail to update when this changes.  Not sure that's a huge deal - any opinion here @aarcro (since you were the original author, so I assume you might be using this blueprint already) @ejholmes @Lowercases @russellballestrini  does remind use this?

File: stacker_blueprints/iam_roles.py [hunk 5]
```diff
@@ -92,17 +147,24 @@ def create_policy(self, name):
...
         )
         self.policies.append(policy)
+        return policy
 
     def create_template(self):
         variables = self.get_variables()
+        created = False
```
Comment (phobologic) [inline stacker_blueprints/iam_roles.py]: You should be able to catch this earlier with something like:

```
if not (variables["Ec2Roles"] or variables["LambdaRoles"]):
    raise InvalidConfig(...)
```

**Comment by dpetzold** [comment]
Hey @phobologic,

Thank you for your feedback. I believe both of the remaining points revolve around the interface to this blueprint which I do not quite understand. Clearing that up will help me with completing this PR.  The current implementation wants to allow for many roles Ec2 and/or Lambda to be specified and attach the same policy to them. I am not sure what the use case for that is.

The case I am trying to handle here is I want one Ec2 role to be created with an Instance Profile that has many managed policies attached. The reason I am asking is that currently by accepting many roles as it does makes changes like this more complicated than it seems they should be.

Typically I would be much inclined to keep an existing interface but I does not seem like it ever worked with policies.

```python
Traceback (most recent call last):
  File "/Users/dpetzold/src/stacker_blueprints/tests/test_iam_roles.py", line 65, in test_lambda_role
    blueprint.create_template()
  File "/Users/dpetzold/src/stacker_blueprints/stacker_blueprints/iam_roles.py", line 108, in create_template
    self.create_policy()
TypeError: create_policy() takes exactly 2 arguments (1 given)
```

https://github.com/cloudtools/stacker_blueprints/commit/2cdf65345759e164857fe5737cd8898808484893#diff-6e6cacc649079d1431dc1c7de1991c80

What I would like to do is something more like this:

```
    VARIABLES = {                                                                                                                         
        "RoleName": {                                                                                                                     
            "type": str,                                                                                                                  
            "description": "The name of the role",                                                                                        
        },                                                                                                                                
        "AttachedPolicies": {                                                                                                             
            "type": list,                                                                                                                 
            "description": "List of ARNs of policies to attach",                                                                          
            "default": [],                                                                                                                
        },                                                                                                                                
        "Path": {                                                                                                                         
            "type": str,                                                                                                                  
            "description": "Provide the path",                                                                                            
            "default": "",                                                                                                                
        },                                                                                                                                
        "InstanceProfile": {                                                                                                              
            "type": bool,                                                                                                                 
            "description": "The role is an instance profile.",                                                                            
            "default": False,                                                                                                             
        },                                                                                                                                
        "Lambda": {                                                                                                                       
            "type": bool,                                                                                                                 
            "description": "This is an lambda role",                                                                                       
            "default": False                                                                                                              
        },                                                                                                                                
    }                                                                                                                                     
```

InstanceProfile and Lambda would be mutually exclusive. I think that would make more sense going forward but again I do not understand the motivation for the current implementation. Please let me know what you think.

Thanks,

**Comment by phobologic** [comment]
Hey @dpetzold - so I think the idea behind this blueprint was to make it so you could easily create a few roles with the same policies.  I think the likelihood that you'd mix one stack config with both ec2 and lambda roles was not likely, but the option was left open.

Honestly, it may be that this blueprint isn't a good fit for exactly what you're working for.  Maybe it's worth creating a specific `EC2Role` blueprint in this same module?  Then you won't be tied to the Lambda stuff here.

**Comment by russellballestrini** [comment]
@dpetzold, Thank you very much for touching this blueprint up, looks good to me.

@phobologic, We do not use this blueprint internally at (Remind)[https://www.remind.com]

@aarcro, What do you think?

**Comment by dpetzold** [comment]
Closed in favor of https://github.com/cloudtools/stacker_blueprints/pull/22
